### PR TITLE
encryption: Fix panic when restore backup when encryption is enabled

### LINKED ISSUE #7485: encryption: tikv panic when encryption is enabled after restore a backup
### What version of TiKV are you using?
tikv master

### Steps to reproduce
turn on encryption on an empty tikv, restore a backup. restore again (or run some write workload), tikv panic with error:
```
[2020/04/16 12:22:54.418 +08:00] [FATAL] [lib.rs:480] ["rocksdb background error. db: kv, reason: compaction, error: Corruption: bad checksum type"]
``` 

### What did you expect?
tikv should restore backup successfully

### What did happened?
tikv panic

### PR DESCRIPTION
**Pull request by yiwu-arbug**
### What problem does this PR solve?

Issue Number: close #7485 

Problem Summary: Encryption was not properly handled in restore flow. Fixing it. See details below:

### What is changed and how it works?

What's Changed:
There are multiple file operations when tikv restore a backup.
1. SST is copied into temp directory from external storage
2. The file is validated using `SstReader`
3. rewrite the SST
3.1 if no key rewritten, the temp SST file is copy to import directory
3.2 If key is rewritten, rewrite the file using `SstWriter`
4. the file is copy again into temp directory so to make the ingestion retriable.
5. ingest the SST into rocksdb

What's fixed in this PR:
* In step 1, wrap the file writer with `EncrypterWriter` to make the local copy encrypted.
* In step 2, pass `EncryptedEnv` into `SstReader` to be able to read the file
* In step 4, call `key_manager.link_file()` so to keep track of the copy. To resolve dependency, move `prepare_sst_for_ingestion` from engine_rocks crate to sst_importer crate.
* Update rocksdb to fix an issue in step 5, where rocksdb call `key_manager.new_file()` where it should call `key_manager.get_file()` to obtain encryption info.

### Related changes
rocksdb fix: https://github.com/tikv/rocksdb/pull/167

### Check List 

Tests 
manual test that #7485 is fixed.

### Release note

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -173,6 +173,9 @@ protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "b6
 fail = { git = "https://github.com/tikv/fail-rs.git", rev = "2cf1175a1a5cc2c70bd20ebd45313afd69b558fc" }
 prometheus = { git = "https://github.com/tikv/rust-prometheus.git", rev = "086b7059482f1b06748e20c9a43efbc78b7252b2" }
 
+[patch."https://github.com/tikv/rust-rocksdb.git"]
+rocksdb = { git = "https://github.com/yiwu-arbug/rust-rocksdb.git", package = "rocksdb", branch = "enc_rwfile" }
+
```
Comment (overvenus) [inline Cargo.toml:178]: Please remove this patch before merging.

File: components/encryption/src/manager/mod.rs [hunk 2]
```diff
@@ -162,21 +164,31 @@ impl Dicts {
         file.iv = Iv::new_ctr().as_slice().to_vec();
         file.key_id = self.key_dict.current_key_id;
         file.method = compat(method);
-        self.file_dict.files.insert(fname.to_owned(), file);
+        self.file_dict.files.insert(fname.to_owned(), file.clone());
```
Comment (overvenus) [inline components/encryption/src/manager/mod.rs:167]: Please do not clone file, `Iv` impls `Copy`.

File: components/sst_importer/src/sst_importer.rs [hunk 3]
```diff
@@ -27,12 +31,17 @@ use crate::metrics::*;
...
 pub struct SSTImporter {
     dir: ImportDir,
+    key_manager: Option<Arc<DataKeyManager>>,
 }
 
 impl SSTImporter {
-    pub fn new<P: AsRef<Path>>(root: P) -> Result<SSTImporter> {
+    pub fn new<P: AsRef<Path>>(
+        root: P,
+        key_manager: Option<Arc<DataKeyManager>>,
```
Comment (overvenus) [inline components/sst_importer/src/sst_importer.rs:40]: We can avoid sst_importer depends on encryption by `add_file` or `link_file` via `EncryptedEnv`, if the change is too large, we can postpone it.

Comment (yiwu-arbug) [reply components/sst_importer/src/sst_importer.rs:40]: I think we need to abstract `Env` out and use it consistently in tikv for file operations. Let's postpone for later.

File: components/encryption/src/manager/mod.rs (hunk 2 above)
Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:167]: The problem is the `file` is a `kvproto::protos::encryptionpb::FileInfo`, which doesn't implement `Copy` trait, and we don't have control of how protobuf generates code.

Comment (overvenus) [reply components/encryption/src/manager/mod.rs:167]: I mean we could just copy IV.

Comment (yiwu-arbug) [reply components/encryption/src/manager/mod.rs:168]: I see. Fixed.

**Pull request merged**